File: sdks/python/examples/offramp_gateway.py
"""Off-ramp via the direct Gateway contract path (method="gateway").

The PaycrestClient accepts a GatewayTransactor — any object implementing
the ``GatewayTransactor`` protocol. This example sketches an adapter on
top of ``web3.py``; you can substitute ``eth-account`` + raw JSON-RPC or
any other wallet library your app already uses.
"""

from __future__ import annotations

import os

from paycrest_sdk import (
    GatewayCreateOrderArgs,
    GatewayPathConfig,
    PaycrestClient,
)


class Web3PyTransactor:
    """Thin web3.py adapter. Construct once per process and reuse."""

    def __init__(self, rpc_url: str, private_key: str):
        from web3 import Web3  # imported lazily — web3 is an optional dep

        self._w3 = Web3(Web3.HTTPProvider(rpc_url))
        self._acct = self._w3.eth.account.from_key(private_key)
        # Minimal ABI fragments lifted from the aggregator.
        self._erc20_abi = [
            {"type": "function", "name": "approve", "stateMutability": "nonpayable",
             "inputs": [{"name": "spender", "type": "address"}, {"name": "amount", "type": "uint256"}],
             "outputs": [{"name": "", "type": "bool"}]},
            {"type": "function", "name": "allowance", "stateMutability": "view",
             "inputs": [{"name": "owner", "type": "address"}, {"name": "spender", "type": "address"}],
             "outputs": [{"name": "", "type": "uint256"}]},
        ]
        self._gateway_abi = [
            {"type": "function", "name": "createOrder", "stateMutability": "nonpayable",
             "inputs": [
                 {"name": "_token", "type": "address"},
                 {"name": "_amount", "type": "uint256"},
                 {"name": "_rate", "type": "uint96"},
                 {"name": "_senderFeeRecipient", "type": "address"},
                 {"name": "_senderFee", "type": "uint256"},
                 {"name": "_refundAddress", "type": "address"},
                 {"name": "messageHash", "type": "string"},
             ],
             "outputs": [{"name": "orderId", "type": "bytes32"}]},
        ]

    def chain_id(self) -> int:
        return self._w3.eth.chain_id

    def sender_address(self) -> str:
        return self._acct.address

    def allowance(self, token: str, owner: str, spender: str) -> int:
        erc20 = self._w3.eth.contract(address=self._w3.to_checksum_address(token), abi=self._erc20_abi)
        return int(erc20.functions.allowance(owner, spender).call())

    def approve(self, token: str, spender: str, amount: int) -> str:
        erc20 = self._w3.eth.contract(address=self._w3.to_checksum_address(token), abi=self._erc20_abi)
        tx = erc20.functions.approve(spender, amount).build_transaction({
            "from": self._acct.address,
            "nonce": self._w3.eth.get_transaction_count(self._acct.address),
        })
        signed = self._w3.eth.account.sign_transaction(tx, self._acct.key)
        tx_hash = self._w3.eth.send_raw_transaction(signed.rawTransaction)
        return tx_hash.hex()

    def create_order(self, gateway: str, args: GatewayCreateOrderArgs) -> str:
        gw = self._w3.eth.contract(address=self._w3.to_checksum_address(gateway), abi=self._gateway_abi)
        tx = gw.functions.createOrder(
            args.token,
            args.amount,
            args.rate,
            args.sender_fee_recipient,
            args.sender_fee,
            args.refund_address,
            args.message_hash,
        ).build_transaction({
            "from": self._acct.address,
            "nonce": self._w3.eth.get_transaction_count(self._acct.address),
        })
        signed = self._w3.eth.account.sign_transaction(tx, self._acct.key)
        tx_hash = self._w3.eth.send_raw_transaction(signed.rawTransaction)
        return tx_hash.hex()


def main() -> None:
    transactor = Web3PyTransactor(
        rpc_url=os.environ["BASE_RPC_URL"],
        private_key=os.environ["SIGNER_PRIVATE_KEY"],
    )
    client = PaycrestClient(
        sender_api_key=os.environ.get("PAYCREST_SENDER_API_KEY"),
        gateway=GatewayPathConfig(transactor=transactor),
    )

    result = client.sender().create_offramp_order(
        {
            "amount": "100",
            "source": {
                "type": "crypto",
                "currency": "USDT",
                "network": "base",
                "refundAddress": transactor.sender_address(),
            },
            "destination": {
                "type": "fiat",
                "currency": "NGN",
                "recipient": {
                    "institution": "GTBINGLA",
                    "accountIdentifier": "1234567890",
                    "accountName": "Jane Doe",
                    "memo": "Payout",
                },
            },
        },
        method="gateway",
    )
    print(result)


if __name__ == "__main__":
    main()

File: sdks/rust/examples/offramp_gateway.rs
//! Off-ramp via the direct Gateway contract path (OfframpMethod::Gateway).
//!
//! The Rust SDK delegates signing + broadcasting to a caller-supplied
//! [`GatewayTransactor`] trait object. The snippet below shows the
//! shape of an ethers-rs adapter — drop it into your integrator and
//! wire up the wallet / provider as usual.

use std::sync::Arc;

use async_trait::async_trait;
use num_bigint::BigUint;
use paycrest_sdk::{
    ClientOptions, GatewayCreateOrderArgs, GatewayPathConfig, GatewayTransactor, OfframpMethod,
    OfframpOrderOutcome, PaycrestClient, PaycrestError,
};
use serde_json::json;

struct DummyTransactor;

#[async_trait]
impl GatewayTransactor for DummyTransactor {
    async fn chain_id(&self) -> Result<u64, PaycrestError> {
        Ok(8453)
    }

    fn from(&self) -> String {
        "0xSenderEoa".to_string()
    }

    async fn allowance(
        &self,
        _token: &str,
        _owner: &str,
        _spender: &str,
    ) -> Result<BigUint, PaycrestError> {
        Ok(BigUint::from(0u64))
    }

    async fn approve(
        &self,
        _token: &str,
        _spender: &str,
        _amount: BigUint,
    ) -> Result<String, PaycrestError> {
        Ok("0xApproveTxHash".to_string())
    }

    async fn create_order(
        &self,
        _gateway: &str,
        _args: GatewayCreateOrderArgs,
    ) -> Result<String, PaycrestError> {
        Ok("0xCreateTxHash".to_string())
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let client = PaycrestClient::new_with_options(ClientOptions {
        sender_api_key: Some("sender-key".to_string()),
        base_url: paycrest_sdk::DEFAULT_BASE_URL.to_string(),
        gateway: Some(GatewayPathConfig {
            transactor: Arc::new(DummyTransactor),
            aggregator_public_key: None,
        }),
        ..Default::default()
    });

    let sender = client.sender()?;
    let payload = json!({
        "amount": "100",
        "rate": "1500",
        "source": {"type": "crypto", "currency": "USDT", "network": "base", "refundAddress": "0xSenderEoa"},
        "destination": {
            "type": "fiat",
            "currency": "NGN",
            "recipient": {
                "institution": "GTBINGLA",
                "accountIdentifier": "1234567890",
                "accountName": "Jane Doe",
                "memo": "Payout"
            }
        }
    });

    match sender.create_offramp_order_with_method(payload, OfframpMethod::Gateway).await? {
        OfframpOrderOutcome::Gateway(result) => println!("{:?}", result),
        OfframpOrderOutcome::Api(_) => unreachable!(),
    }
    Ok(())
}
